fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt

```diff
@@ -302,6 +314,20 @@ class WebExtensionPromptFeature(
         consumePromptRequest()
     }
 
+    private fun handleApprovedOptionalPermissions(
+        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
+    ) {
+        promptRequest.onConfirm(true)
+        consumePromptRequest()
+    }
+
+    private fun handleDeniedOptionalPermissions(
+        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
+    ) {
+        promptRequest.onConfirm(false)
+        consumePromptRequest()
+    }
+
```

Comment:
Noticed we had some duplication between the permissions handling, as they don't belong to the same object hierarchy, it was a bit cumbersome to handle. Below I added a diff, where I put them into the same hierarchy reducing the duplicated code, and make it a bit clear to work with them. What do you think?   


```diff
diff --git a/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt b/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt
index 87152b3c42..c22c2cab3f 100644
--- a/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt
+++ b/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/extension/WebExtensionPromptRequest.kt
@@ -38,27 +38,34 @@ sealed class WebExtensionPromptRequest {
      * @param extension The [WebExtension] that requested the dialog to be shown.
      */
     sealed class AfterInstallation(open val extension: WebExtension) : WebExtensionPromptRequest() {
-        /**
-         * Value type that represents a request for a permission prompt.
-         * @property extension The [WebExtension] that requested the dialog to be shown.
-         * @property onConfirm A callback indicating whether the permissions were granted or not.
-         */
-        data class Permissions(
-            override val extension: WebExtension,
-            val onConfirm: (Boolean) -> Unit,
-        ) : AfterInstallation(extension)
 
-        /**
-         * Value type that represents a request for a (optional) permission prompt.
-         * @property extension The [WebExtension] that requested the dialog to be shown.
-         * @property permissions The optional permissions to list in the dialog.
-         * @property onConfirm A callback indicating whether the permissions were granted or not.
-         */
-        data class OptionalPermissions(
+
+        sealed class Permissions(
             override val extension: WebExtension,
-            val permissions: List<String>,
-            val onConfirm: (Boolean) -> Unit,
-        ) : AfterInstallation(extension)
+            open val onConfirm: (Boolean) -> Unit,
+        ) : AfterInstallation(extension) {
+            /**
+             * Value type that represents a request for a permission prompt.
+             * @property extension The [WebExtension] that requested the dialog to be shown.
+             * @property onConfirm A callback indicating whether the permissions were granted or not.
+             */
+            data class Required(
+                override val extension: WebExtension,
+                override val onConfirm: (Boolean) -> Unit,
+            ) : Permissions(extension, onConfirm)
+
+            /**
+             * Value type that represents a request for a (optional) permission prompt.
+             * @property extension The [WebExtension] that requested the dialog to be shown.
+             * @property permissions The optional permissions to list in the dialog.
+             * @property onConfirm A callback indicating whether the permissions were granted or not.
+             */
+            data class Optional(
+                override val extension: WebExtension,
+                val permissions: List<String>,
+                override val onConfirm: (Boolean) -> Unit,
+            ) : Permissions(extension, onConfirm)
+        }
 
         /**
          * Value type that represents a request for showing post-installation prompt.
diff --git a/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt b/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt
index 674d5734d3..86f99ca929 100644
--- a/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt
+++ b/fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt
@@ -77,11 +77,11 @@ class WebExtensionPromptFeature(
         // opens the add-ons manager.
         val addon = Addon.newFromWebExtension(promptRequest.extension)
         when (promptRequest) {
-            is WebExtensionPromptRequest.AfterInstallation.Permissions -> handlePermissionRequest(
+            is WebExtensionPromptRequest.AfterInstallation.Permissions.Required -> handleRequiredPermissionRequest(
                 addon,
                 promptRequest,
             )
-            is WebExtensionPromptRequest.AfterInstallation.OptionalPermissions -> handleOptionalPermissionsRequest(
+            is WebExtensionPromptRequest.AfterInstallation.Permissions.Optional -> handleOptionalPermissionsRequest(
                 addon,
                 promptRequest,
             )
@@ -108,34 +108,34 @@ class WebExtensionPromptFeature(
         showPostInstallationDialog(addon)
     }
 
-    private fun handlePermissionRequest(
+    private fun handleRequiredPermissionRequest(
         addon: Addon,
         promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions,
     ) {
         showPermissionDialog(
             addon = addon,
-            onPositiveButtonClicked = { handleApprovedPermissions(promptRequest) },
-            onNegativeButtonClicked = { handleDeniedPermissions(promptRequest) },
+            onPositiveButtonClicked = { handlePermissions(promptRequest, granted = true) },
+            onNegativeButtonClicked = { handlePermissions(promptRequest, granted = true) },
         )
     }
 
     @VisibleForTesting
     internal fun handleOptionalPermissionsRequest(
         addon: Addon,
-        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
+        promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions.Optional,
     ) {
         val shouldGrantWithoutPrompt = Addon.localizePermissions(promptRequest.permissions, context).isEmpty()
 
         // If we don't have any promptable permissions, just proceed.
         if (shouldGrantWithoutPrompt) {
-            handleApprovedOptionalPermissions(promptRequest)
+            handlePermissions(promptRequest, granted = true)
             return
         }
 
         showPermissionDialog(
             addon = addon,
-            onPositiveButtonClicked = { handleApprovedOptionalPermissions(promptRequest) },
-            onNegativeButtonClicked = { handleDeniedOptionalPermissions(promptRequest) },
+            onPositiveButtonClicked = { handlePermissions(promptRequest, granted = true) },
+            onNegativeButtonClicked = { handlePermissions(promptRequest, granted = true) },
             forOptionalPermissions = true,
             optionalPermissions = promptRequest.permissions,
         )
@@ -252,31 +252,15 @@ class WebExtensionPromptFeature(
         findPreviousPermissionDialogFragment()?.let { dialog ->
             dialog.onPositiveButtonClicked = { addon ->
                 store.state.webExtensionPromptRequest?.let { promptRequest ->
-                    when (promptRequest) {
-                        is WebExtensionPromptRequest.AfterInstallation.Permissions -> {
-                            if (addon.id == promptRequest.extension.id) {
-                                handleApprovedPermissions(promptRequest)
-                            }
-                        }
-                        is WebExtensionPromptRequest.AfterInstallation.OptionalPermissions -> {
-                            if (addon.id == promptRequest.extension.id) {
-                                handleApprovedOptionalPermissions(promptRequest)
-                            }
-                        }
-                        else -> { /* nothing to do for the other requests. */ }
+                    if (promptRequest is WebExtensionPromptRequest.AfterInstallation.Permissions && addon.id == promptRequest.extension.id) {
+                        handlePermissions(promptRequest, granted = true)
                     }
                 }
             }
             dialog.onNegativeButtonClicked = {
                 store.state.webExtensionPromptRequest?.let { promptRequest ->
-                    when (promptRequest) {
-                        is WebExtensionPromptRequest.AfterInstallation.Permissions -> {
-                            handleDeniedPermissions(promptRequest)
-                        }
-                        is WebExtensionPromptRequest.AfterInstallation.OptionalPermissions -> {
-                            handleDeniedOptionalPermissions(promptRequest)
-                        }
-                        else -> { /* nothing to do for the other requests. */ }
+                    if (promptRequest is WebExtensionPromptRequest.AfterInstallation.Permissions) {
+                        handlePermissions(promptRequest, granted = false)
                     }
                 }
             }
@@ -304,27 +288,9 @@ class WebExtensionPromptFeature(
         }
     }
 
-    private fun handleDeniedPermissions(promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions) {
-        promptRequest.onConfirm(false)
-        consumePromptRequest()
-    }
-
-    private fun handleApprovedPermissions(promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions) {
-        promptRequest.onConfirm(true)
-        consumePromptRequest()
-    }
 
-    private fun handleApprovedOptionalPermissions(
-        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
-    ) {
-        promptRequest.onConfirm(true)
-        consumePromptRequest()
-    }
-
-    private fun handleDeniedOptionalPermissions(
-        promptRequest: WebExtensionPromptRequest.AfterInstallation.OptionalPermissions,
-    ) {
-        promptRequest.onConfirm(false)
+    private fun handlePermissions(promptRequest: WebExtensionPromptRequest.AfterInstallation.Permissions, granted: Boolean) {
+        promptRequest.onConfirm(granted)
         consumePromptRequest()
     }
 ```

Reply:
loevely, thanks!